src/core/tsi/s2a/handshaker/s2a_handshaker_client.cc

```diff
@@ -175,18 +181,34 @@ tsi_result S2AHandshakerClient::Next(grpc_slice* bytes_received) {
   }
   grpc_byte_buffer_destroy(send_buffer_);
   send_buffer_ = buffer;
-  tsi_result call_result = MakeGrpcCall(/*is_start=*/false);
-  if (call_result != TSI_OK) {
-    gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+  if (!no_calls_for_testing_) {
+    tsi_result call_result = MakeGrpcCall(/* is_start=*/false);
+    if (call_result != TSI_OK) {
+      gpr_log(GPR_ERROR, kS2AMakeGrpcCallFailed);
+      return call_result;
+    }
   }
-  return call_result;
+  return TSI_OK;
 }
 
 /** ------------------- Callback methods. -------------------------------- **/
 
 static void S2AOnStatusReceived(void* arg, grpc_error* error) {
-  // TODO(mattstev): implement.
-  return;
+  S2AHandshakerClient* client = static_cast<S2AHandshakerClient*>(arg);
+  GPR_ASSERT(client != nullptr);
+  if (client->handshake_status_code() != GRPC_STATUS_OK) {
+    char* status_details =
+        grpc_slice_to_c_string(client->handshake_status_details());
+    gpr_log(GPR_INFO,
+            "S2AHandshakerClient:%p on_status_received status:%d "
+            "details:|%s| error:|%s|",
+            client, client->handshake_status_code(), status_details,
+            grpc_error_string(error));
+    gpr_free(status_details);
+  }
+  client->MaybeCompleteTsiNext(/* receive_status_finished=*/true,
+                               /* pending_recv_message_result=*/nullptr);
```

Comment:
Don't we also need a global queue (similar to ALTS) to handle a large number of concurrent connection requests?

Reply:
As we discussed offline, we do not need the global queue. Limiting the number of concurrent connection requests will be handled by the S2A service.

Reply:
Ack.